### PR TITLE
rpi led and serial

### PR DESCRIPTION
**Pull request by markfirmware**
zig build -Dtarget=aarch64-freestanding

Then copy rpi-sdcard.img to an unused sd card and insert in 3b+

Will flash rpi3b+ activity led quickly on exception. Pins 14/15 have serial tx/rx.

### CODE REVIEW

File: build.zig [hunk 1]
```diff
@@ -160,4 +160,49 @@ pub fn build(b: *Builder) !void {
         "target remote localhost:1234",
     });
     debug_step.dependOn(&debug_cmd.step);
+
+    const zip_folder = "rpi-sdcard";
```
Comment (SamTebbs33) [inline build.zig]: Instead of zipping, could a .img file not be created that has the kernel, config file et.c inside it? Then that could be copied directly to an SD card instead of having to unzip.

File: build.zig [hunk 2]
```diff
@@ -160,4 +160,49 @@ pub fn build(b: *Builder) !void {
...
+        firmware_url ++ "start.elf",
+    });
+    const zip_do_zip_cmd = b.addSystemCommand(&[_][]const u8{
+        "zip",
+        "-jqr",
+        zip_folder ++ ".zip",
+        zip_folder,
+    });
+
+    const zip_step = b.step("zip", "Create product zip file");
```
Comment (SamTebbs33) [inline build.zig]: I think it would be best to do all of this under the default step, rather than in a specialised zip step.

File: build.zig (hunk 1 above)
Comment (SamTebbs33) [inline build.zig]: I think it would be cleaner to put the build zip/image file into the output directory (`b.cache_root`). This will keep built files in one place.

File: src/kernel/arch/aarch64/arch.zig [hunk 3]
```diff
@@ -35,50 +36,38 @@ pub fn initTTY(boot_payload: BootPayload) TTY {
...
+        mmio.write(mmio_addr, .AUX_ENABLES, 1);
+        mmio.write(mmio_addr, .AUX_MU_IER_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_CNTL_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_LCR_REG, 3);
+        mmio.write(mmio_addr, .AUX_MU_MCR_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_IER_REG, 0);
+        mmio.write(mmio_addr, .AUX_MU_IIR_REG, 0xc6);
+        mmio.write(mmio_addr, .AUX_MU_BAUD_REG, 270);
+        mmio.write(mmio_addr, .AUX_MU_CNTL_REG, 3);
+    }
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/arch.zig]: Why did all of this need to be changed? All the comments, system clock configuration and such has been removed.

File: src/kernel/arch/aarch64/arch.zig [hunk 4]
```diff
@@ -35,50 +36,38 @@ pub fn initTTY(boot_payload: BootPayload) TTY {
...
 }
 
-fn uartWriteByte(byte: u8) void {
-    // Wait until the UART is ready to transmit
-    while ((mmio.read(mmio_addr, mmio.Register.UART_FLAGS) & (1 << 5)) != 0) {}
-    mmio.write(mmio_addr, mmio.Register.UART_DATA, byte);
+pub fn uartWriteByte(byte: u8) void {
+    if (byte == 10) {
+        uartWriteByte(13);
+    }
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/arch.zig:72]: What's the purpose of this check?

File: src/kernel/arch/aarch64/arch.zig [hunk 5]
```diff
@@ -120,5 +109,133 @@ pub fn haltNoInterrupts() noreturn {
 
 // TODO: implement
 pub fn spinWait() noreturn {
-    while (true) {}
+    spinLed(500);
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/arch.zig]: I think this should still be a while true loop (it should probably include a `wfe` but that can wait) since you won't always want to make the LED flash when spinning.

File: src/kernel/arch/aarch64/arch.zig [hunk 6]
```diff
@@ -120,5 +109,133 @@ pub fn haltNoInterrupts() noreturn {
...
+        \\ mov x2, #0x08000000
+        \\ orr x1, x1, x2
+        \\ str w1, [x0, #0x08]
+        \\ mov x1, #0x20000000
+        \\ str w1, [x0, #0x28]
+        :
+        :
+        : "x0", "x1", "x2"
+    );
+}
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/arch.zig]: Could these use more zig code instead? We like to keep raw assembly to the minimum.

File: src/kernel/arch/aarch64/arch.zig [hunk 7]
```diff
@@ -120,5 +109,133 @@ pub fn haltNoInterrupts() noreturn {
...
+    AlternateFunction3 = 7,
+    AlternateFunction4 = 3,
+    AlternateFunction5 = 2,
+};
+
+const PinPull = enum {
+    None,
+    Down,
+    Up,
+};
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/arch.zig]: The raspberry pi-specific things (like LEDs and pins) should be moved to the rpi.zig file since they are specific to the rpi.

File: src/kernel/arch/aarch64/arch.zig [hunk 8]
```diff
@@ -120,5 +109,133 @@ pub fn haltNoInterrupts() noreturn {
...
+        \\ msr ttbr0_el3, x1
+        \\ isb
+        \\ mrs x0, sctlr_el3
+        \\ orr x0, x0, #1
+        \\ msr sctlr_el3, x0
+        \\ isb
+        :
+        :
+        : "x0", "x1", "x2", "x3"
+    );
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/arch.zig]: Same again regarding the raw assembly code.

File: src/kernel/arch/aarch64/boot.zig [hunk 9]
```diff
@@ -8,32 +9,39 @@ extern var KERNEL_STACK_END: *usize;
...
+    asm volatile (
+        \\ mrs x0, mpidr_el1
+        \\ mov x1, #3
+        \\ ands x0, x0, x1
+        \\ beq cpu0
+        \\hang:
+        \\ wfe
+        \\ b hang
+        \\cpu0:
+    );
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/boot.zig]: Raw assembly again :)

File: src/kernel/arch/aarch64/boot.zig [hunk 10]
```diff
@@ -8,32 +9,39 @@ extern var KERNEL_STACK_END: *usize;
...
+        \\ msr vbar_el3, %[table_addr]
         :
         : [table_addr] "r" (@ptrToInt(&interrupts.exception_table))
     );
 
+    arch.enableFlatMmu();
+
     // The rpi puts the board part number in midr_el1
-    const board = detectBoard();
+    const board = rpi.RaspberryPiBoard.RaspberryPi3; // detectBoard();
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/boot.zig]: Why not `detectBoard()`?

File: src/kernel/arch/aarch64/mmio.zig [hunk 11]
```diff
@@ -1,31 +1,87 @@
-const GPIO_OFFSET: usize = 0x200000;
-const UART_OFFSET: usize = GPIO_OFFSET + 0x1000;
+const log = @import("../../log.zig");
+const std = @import("std");
+
+const MMIO_OFFSET: usize = 0x200000;
+const GPIO_OFFSET: usize = MMIO_OFFSET + 0x0000;
+const UART_OFFSET: usize = MMIO_OFFSET + 0x1000;
+const AUX_OFFSET: usize = MMIO_OFFSET + 0x15000;
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/mmio.zig:7]: Nice. Making them relative to mmio_offset is more correct.

File: src/kernel/arch/aarch64/mmio.zig [hunk 12]
```diff
@@ -1,31 +1,87 @@
...
+    GPIO_CLR0 = GPIO_OFFSET + 0x28,
+    GPIO_CLR1 = GPIO_OFFSET + 0x2c,
+    GPIO_PUD = GPIO_OFFSET + 0x94,
+    GPIO_PUDCLK0 = GPIO_OFFSET + 0x98,
+    GPIO_PUDCLK1 = GPIO_OFFSET + 0x9c,
     MBOX_STATUS = MBOX_OFFSET + 0x18,
     MBOX_WRITE = MBOX_OFFSET + 0x20,
     MBOX_READ = MBOX_OFFSET + 0x00,
+    UART_DR = UART_OFFSET + 0x00,
+    UART_FR = UART_OFFSET + 0x18,
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/mmio.zig]: For the registers that existed before, could we keep old names? They're a bit more descriptive.

File: src/kernel/arch/aarch64/mmio.zig [hunk 13]
```diff
@@ -1,31 +1,87 @@
...
 };
 
 pub fn write(mmio_base: usize, reg: Register, data: u32) void {
     const addr = mmio_base + @enumToInt(reg);
-    const ptr = @intToPtr(*u32, addr);
+    const ptr = @intToPtr(*volatile u32, addr);
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/mmio.zig:52]: Good spot :+1:

**Review by SamTebbs33 — CHANGES_REQUESTED** [review]
Good job on getting it working in hardware! You're not going to like me for requesting this but could you add doc comments above each function, struct, struct field and public variable? Documentation is important :) Have a look at src/kernel/filesystem/vfs.zig for inspiration.

File: build.zig (hunk 2 above)
Comment (SamTebbs33) [reply build.zig]: Of course these zipping commands and such shouldn't be added when building for x86 so a check for aarch64 would be needed too.

File: src/kernel/arch/aarch64/arch.zig (hunk 4 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig:72]: This sends a carriage return before a line feed. Should I add a comment or is there a better place to do this?

File: src/kernel/arch/aarch64/arch.zig (hunk 6 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig]: Yes. I wanted to get your eyes on this before proceeding further. Updating tcr and ttbr0 is delicate and will probably still be asm copied from the arm manual.

File: src/kernel/arch/aarch64/arch.zig (hunk 7 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig]: Done: b86e4cdb62bf8cd9c6ebd743986baca6d3e839ce

File: src/kernel/arch/aarch64/boot.zig (hunk 9 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/boot.zig]: This is necessary because the disassembly of the zig version showed the use of the sp which is not yet initialized. Each core will need its own sp so sp cannot be initialized yet.

I have added a comment: a161e9f4352fd8f183017710984af0c6335c1e79

File: src/kernel/arch/aarch64/boot.zig (hunk 10 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/boot.zig]: ~~I suspect a compiler defect when returning an optional struct from a function. I intend to reduce and isolate this when I am at my lab today.~~

67dfa7a7e650cb8295daee71b9337a082df720f5

Turns out the problem was returning the address of a const on the stack.

File: src/kernel/arch/aarch64/mmio.zig (hunk 13 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/mmio.zig:52]: Thanks.

File: build.zig (hunk 1 above)
Comment (markfirmware) [reply build.zig]: It can be done but would still require something like https://www.balena.io/etcher/ to do the copying.

Could we open an issue to improve this later?

Comment (markfirmware) [reply build.zig]: > Good job on getting it working in hardware! You're not going to like me for requesting this but could you add doc comments above each function, struct, struct field and public variable? Documentation is important :) Have a look at src/kernel/filesystem/vfs.zig for inspiration.

Will do.

File: src/kernel/arch/aarch64/mmio.zig (hunk 11 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/mmio.zig:7]: Thanks.

File: src/kernel/arch/aarch64/arch.zig (hunk 5 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig]: Done: 313b85716e147c5a5393fb938d40fedf4932d8f4

File: src/kernel/arch/aarch64/mmio.zig (hunk 12 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/mmio.zig]: The reference for these is https://www.raspberrypi.org/documentation/hardware/raspberrypi/bcm2835/BCM2835-ARM-Peripherals.pdf which also applies generally to the bcm2837 which is on the pi3b and pi3b+.

The names I used are from this manual but using your prefix (GP instead of GPIO.) What I would like to do is use the ones straight from the manual, unmodified. They are not as descriptive, but they are what everyone else is using.

Your decision.

File: src/kernel/arch/aarch64/arch.zig (hunk 3 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig]: Uart0 is the full uart and uart1 is the mini (aux) uart. Uart0 is used for the bluetooth chip on 3b and above, leaving uart1 for anything else like the serial console.

However, qemu does not implement uart1, it only implements uart0, and of course it does not implement the bluetooth chip.

Therefore uart1 is used for the console except when under qemu. Under qemu uart0 is used for the console. Also qemu uart0 does not require initialization so uart0 constants were removed. Uart0 constants and initialization would be added back, probably differently, for bluetooth when we get to it.

By the way, the uart0 interface to the bluetooth chip is quite elegant and easy to program.

File: src/kernel/arch/aarch64/arch.zig (hunk 6 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig]: These are the instructions for updating tcr and ttbr0. This is the 1.0 manual. There is a 1.1 manual but it does not have an online link - the 1.1 version can only be used locally. This particukar section of the manual is the same.

https://static.docs.arm.com/100940/0100/armv8_a_address%20translation_100940_0100_en.pdf#%5B%7B%22num%22%3A53%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C54%2C456%2C0%5D

**Review by markfirmware — COMMENTED** [review]
Some of the comments can be resolved. @SamTebbs33

File: build.zig (hunk 1 above)
Comment (SamTebbs33) [reply build.zig]: Oh I don't mean that the SD card should be flashed with this step, just that an image file is created (that can be put on an SD card manually), just like we do with `grub-mkrescue` for x86.

File: src/kernel/arch/aarch64/arch.zig (hunk 3 above)
Comment (SamTebbs33) [reply src/kernel/arch/aarch64/arch.zig]: I see. So does uart1 not depend on the system clock? This code could do with some comments to explain what is happening.

File: src/kernel/arch/aarch64/arch.zig (hunk 4 above)
Comment (SamTebbs33) [reply src/kernel/arch/aarch64/arch.zig:72]: Hmm sending a carriage return as well sounds very windows-like. Is this standard in serial communication?

File: src/kernel/arch/aarch64/arch.zig (hunk 6 above)
Comment (SamTebbs33) [reply src/kernel/arch/aarch64/arch.zig]: > Yes. I wanted to get your eyes on this before proceeding further. Updating tcr and ttbr0 is delicate and will probably still be asm copied from the arm manual.

Sure, if something has to use inline assembly that's fine. Register moves, arithmetic and the like should preferably use Zig.

File: src/kernel/arch/aarch64/boot.zig (hunk 9 above)
Comment (SamTebbs33) [reply src/kernel/arch/aarch64/boot.zig]: Oh well spotted. It's stupid use of the stack too since the stored value is loaded immediately afterwards. Perhaps you could read `mpidr_el1` in assembly then do the comparison with 0 and loop in Zig?

File: src/kernel/arch/aarch64/boot.zig (hunk 10 above)
Comment (SamTebbs33) [reply src/kernel/arch/aarch64/boot.zig]: Nice!

File: src/kernel/arch/aarch64/mmio.zig (hunk 12 above)
Comment (SamTebbs33) [reply src/kernel/arch/aarch64/mmio.zig]: I see your reasoning. I still think that more descriptive names are better though. Perhaps we can convince the rest of the world too :)

**Review by SamTebbs33 — CHANGES_REQUESTED** [review]

File: src/kernel/arch/aarch64/arch.zig (hunk 4 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig:72]: Somewhere along the line a carriage return is needed for the terminal to move the print head to the left position. Line feed only advances the paper. This is real teletype stuff.

This is needed for an attached tty or a terminal emulator like minicom.

File: src/kernel/arch/aarch64/boot.zig (hunk 9 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/boot.zig]: Possibly, but that loop might be using the stack.

Can we keep it assembly in the pr and have an issue to improve it?

File: src/kernel/arch/aarch64/mmio.zig (hunk 12 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/mmio.zig]: Ok, the hex offsets are an unambiguous link between the source and the manual, so more descriptive names will work.

File: src/kernel/arch/aarch64/arch.zig (hunk 4 above)
Comment (SamTebbs33) [reply src/kernel/arch/aarch64/arch.zig:72]: OK that makes sense. You could add a brief comment to explain that.

Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig:72]: Done: 43da14b643f868eb9a12dc80e4c4ddcb8cd21117

File: src/kernel/arch/aarch64/boot.zig (hunk 9 above)
Comment (SamTebbs33) [reply src/kernel/arch/aarch64/boot.zig]: Have a look and see if looping does end up using the stack in the debug build mode. If it does then I reckon you can keep it as it is.

File: build.zig (hunk 1 above)
Comment (markfirmware) [reply build.zig]: https://forums.balena.io/t/how-do-i-create-an-sd-disk-image-file/171673

File: src/kernel/arch/aarch64/boot.zig (hunk 9 above)
Comment (FireFox317) [reply src/kernel/arch/aarch64/boot.zig]: I think an idea might be to keep the `_start` function really small and only setup a temporary stack (with asm volatile) and call another function. That way the `_start` function will not use the stack and the next function can do the checking of the core number etc and afterwards setting up the correct stack. The small temporary stack only has to be like 256 bytes or something small.

Or just setup a small temporary stack as the first step in the `_start` function.

Comment (SamTebbs33) [reply src/kernel/arch/aarch64/boot.zig]: > I think an idea might be to keep the `_start` function really small and only setup a temporary stack (with asm volatile) and call another function. That way the `_start` function will not use the stack and the next function can do the checking of the core number etc and afterwards setting up the correct stack. The small temporary stack only has to be like 256 bytes or something small.
> 
> Or just setup a small temporary stack as the first step in the `_start` function.

@FireFox317 That sounds ideal to me. The only consideration would be if the cores using the same stack would cause a conflict.

File: src/kernel/arch/aarch64/rpi.zig [hunk 14]
```diff
@@ -47,6 +48,7 @@ pub inline fn turnOnLed() void {
 }
 
 pub inline fn turnOffLed() void {
+    // do not use zig here in order to side-step occasional compiler defects
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/rpi.zig]: What compiler defects have you encountered in `turnOffLed` and `turnOnLed`?

File: src/kernel/arch/aarch64/boot.zig [hunk 15]
```diff
@@ -20,6 +20,9 @@ export fn _start() linksection(".text.boot") callconv(.Naked) noreturn {
         \\ wfe
         \\ b hang
         \\cpu0:
+        \\ mov x0, #0x3f200000
+        \\ mov w1, #67
+        \\ str w1, [x0, 0x1000]
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/boot.zig]: What is this doing? It's also best to avoid using unclear constants.

**Review by SamTebbs33 — CHANGES_REQUESTED** [review]

File: build.zig (hunk 1 above)
Comment (SamTebbs33) [reply build.zig]: I believe that something like this would work https://unix.stackexchange.com/questions/90793/create-iso-image-from-folder-via-terminal-commands/90794

File: src/kernel/arch/aarch64/boot.zig (hunk 9 above)
Comment (FireFox317) [reply src/kernel/arch/aarch64/boot.zig]: @SamTebbs33 yeah obviously you want different stacks per core, otherwise the cores would overwrite the data of the other core. I would do something like the following:
```asm
mrs x0, mpidr_el1 // get system register
and x0, x0, #3 // mask out bits not belonging to core id
mul x1, x0, #255 // 255 bytes per stack
mov x2, <stack_address> // store the lowest stack address in x2
sub x2, x2, x1 
mov sp, x2
```
This can probably written in smaller assembly.

File: src/kernel/arch/aarch64/boot.zig (hunk 15 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/boot.zig]: This was residue from development. I have removed it.

File: src/kernel/arch/aarch64/mmio.zig (hunk 12 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/mmio.zig]: > For the registers that existed before, could we keep old names? They're a bit more descriptive.

Done: 0a034d3c956f98a70001b24ed4453ccb3520fe9d

File: src/kernel/arch/aarch64/boot.zig (hunk 9 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/boot.zig]: @SamTebbs33 I will change to this approach.

Comment (markfirmware) [reply src/kernel/arch/aarch64/boot.zig]: @SamTebbs33 @FireFox317

boot.zig: 

https://github.com/markfirmware/pluto/blob/1b04bd80a27c560a7777542e38d07f2844ff8ecd/src/kernel/arch/aarch64/boot.zig#L1-L50

cpu registers:

https://github.com/markfirmware/pluto/blob/1b04bd80a27c560a7777542e38d07f2844ff8ecd/src/kernel/arch/aarch64/arch.zig#L118-L178

File: src/kernel/arch/aarch64/arch.zig [hunk 16]
```diff
@@ -113,6 +115,74 @@ pub fn spinWait() noreturn {
     while (true) {}
 }
 
+pub const cpu = struct {
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/arch.zig]: This is really nice. We could consider using this in x86 in the future! Don't forget to capitalise `Cpu`.

File: src/kernel/arch/aarch64/boot.zig [hunk 17]
```diff
@@ -1,49 +1,57 @@
...
+            cpu.wfe();
+        }
+    }
+
+    // Give all exception levels the same vector table
+    cpu.vbar.el(1).write(@ptrToInt(&interrupts.exception_table));
+    cpu.vbar.el(2).write(@ptrToInt(&interrupts.exception_table));
+    cpu.vbar.el(3).write(@ptrToInt(&interrupts.exception_table));
+
+    interrupts.exception_handler_depth = 0;
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/boot.zig]: Very nice

File: src/kernel/arch/aarch64/boot.zig [hunk 18]
```diff
@@ -1,49 +1,57 @@
...
 
     arch.enableFlatMmu();
 
-    // The rpi puts the board part number in midr_el1
     board = detectBoard();
+    arch.initMmioAddress(&board);
+
+    // invoke alignment exception
+    // var addr: usize = 1;
+    // @intToPtr(*volatile u32, addr).* = 0;
```
Comment (SamTebbs33) [inline src/kernel/arch/aarch64/boot.zig]: This looks like it's for testing so remember to remove it before we merge.

**Review by SamTebbs33 — CHANGES_REQUESTED** [review]

File: src/kernel/arch/aarch64/boot.zig (hunk 9 above)
Comment (SamTebbs33) [reply src/kernel/arch/aarch64/boot.zig]: Cheers

**Comment by SamTebbs33** [comment]
Other then the comments from my latest review, it looks like all that is needed is some documentation and comments in `initSerial`.

File: src/kernel/arch/aarch64/boot.zig (hunk 18 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/boot.zig]: Done.

File: src/kernel/arch/aarch64/boot.zig (hunk 17 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/boot.zig]: Thanks

File: src/kernel/arch/aarch64/arch.zig (hunk 16 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig]: Thanks. Done.

File: src/kernel/arch/aarch64/arch.zig (hunk 8 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig]: This is no longer assembly but the constants still need explanation.

**Comment by markfirmware** [comment]
> Other then the comments from my latest review, it looks like all that is needed is some documentation and comments in `initSerial`.

Yes including a few questions that you made that I have not yet answered.

File: build.zig (hunk 1 above)
Comment (markfirmware) [reply build.zig]: I took a look but it did not look like a complete solution.

File: build.zig (hunk 2 above)
Comment (markfirmware) [reply build.zig]: Done: https://github.com/markfirmware/pluto/blob/da14037897a64676ef5fcc4d63e23a102690f842/build.zig#L70-L85

File: build.zig (hunk 1 above)
Comment (markfirmware) [reply build.zig]: Done: https://github.com/markfirmware/pluto/blob/da14037897a64676ef5fcc4d63e23a102690f842/build.zig#L70

File: build.zig [hunk 19]
```diff
@@ -67,9 +67,20 @@ pub fn build(b: *Builder) !void {
     exec.setLinkerScriptPath(linker_script_path);
     exec.setTarget(target);
 
+    const zip_folder = try fs.path.join(b.allocator, &[_][]const u8{ b.cache_root, "rpi-sdcard" });
+    const kernel = try fs.path.join(b.allocator, &[_][]const u8{ zip_folder, "kernel8.img" });
+    const firmware_version_tag = "1.20200601+arm64";
+    const firmware_url = "https://github.com/raspberrypi/firmware/raw/" ++ firmware_version_tag ++ "/boot/";
+    const image_file_name = try fs.path.join(b.allocator, &[_][]const u8{ exec.output_dir.?, "kernel8.img" });
```
Comment (SamTebbs33) [inline build.zig]: Can these be declared inside the .aarch64 branch of the switch?

File: build.zig [hunk 20]
```diff
@@ -67,9 +67,20 @@ pub fn build(b: *Builder) !void {
...
     const make_iso = switch (target.getCpuArch()) {
         .i386 => b.addSystemCommand(&[_][]const u8{ "./makeiso.sh", boot_path, modules_path, iso_dir_path, exec.getOutputPath(), output_iso }),
-        .aarch64 => b.addSystemCommand(&[_][]const u8{ "aarch64-linux-gnu-objcopy", exec.getOutputPath(), "-O", "binary", try fs.path.join(b.allocator, &[_][]const u8{ exec.output_dir.?, "kernel8.img" }) }),
+        .aarch64 =>
+        // armstub not yet working, therefore: b 0x80000 (which is 0x14020000) followed by 0 filler until 0x80000 followed by kernel that starts at 0x80000
```
Comment (SamTebbs33) [inline build.zig]: What's wrong with armstub?

File: build.zig [hunk 21]
```diff
@@ -67,9 +67,20 @@ pub fn build(b: *Builder) !void {
...
+        b.addSystemCommand(&[_][]const u8{
+            "/bin/bash",
+            "-c",
+            try std.fmt.allocPrint(b.allocator, "aarch64-linux-gnu-objcopy {} -O binary {} && mkdir --parents {} && cp --archive src/kernel/arch/aarch64/rpi-sdcard/config.txt {} && echo -ne \"\\x00\\x00\\x02\\x14\" > {} && dd status=none bs=1 count=524284 if=/dev/zero >> {} && cat {}/kernel8.img >> {} && wget --directory-prefix={} --quiet --timestamp {}/bootcode.bin {}/fixup.dat {}/start.elf && zip --junk-paths --quiet --recurse-paths {}.zip {}", .{ exec.getOutputPath(), image_file_name, zip_folder, zip_folder, kernel, kernel, b.cache_root, kernel, zip_folder, firmware_url, firmware_url, firmware_url, zip_folder, zip_folder }),
```
Comment (SamTebbs33) [inline build.zig]: It would be cleaner for these to be separate system commands that each depends on the previous one.

**Review by SamTebbs33 — CHANGES_REQUESTED** [review]

File: build.zig (hunk 19 above)
Comment (markfirmware) [reply build.zig]: Done: https://github.com/markfirmware/pluto/blob/da9b1aaef9ef6261e1f6d4001b510201e248a44c/build.zig#L72-L77

@SamTebbs33

File: build.zig (hunk 21 above)
Comment (markfirmware) [reply build.zig]: Done: https://github.com/markfirmware/pluto/blob/da9b1aaef9ef6261e1f6d4001b510201e248a44c/build.zig#L72-L142

@SamTebbs33

File: build.zig (hunk 20 above)
Comment (markfirmware) [reply build.zig]: I do not know. I was unable to get it to work. I can open  an issue to improve this.

@SamTebbs33

File: build.zig [hunk 22]
```diff
@@ -67,20 +67,79 @@ pub fn build(b: *Builder) !void {
...
+            const cp_config = b.addSystemCommand(&[_][]const u8{
+                "cp",
+                "--archive",
+                "src/kernel/arch/aarch64/rpi-sdcard/config.txt",
+                zip_folder,
+            });
+            cp_config.step.dependOn(&objcopy.step);
+
+            const emulate_armstub = armstubSequence: {
+                // armstub not yet working, therefore: b 0x80000 (which is 0x14020000)
```
Comment (SamTebbs33) [inline build.zig]: What's wrong with armstub?

File: build.zig [hunk 23]
```diff
@@ -67,20 +67,79 @@ pub fn build(b: *Builder) !void {
...
+                const create_b_0x80000 = bash(b, "echo -ne \"\\x00\\x00\\x02\\x14\" > {}", .{kernel});
+                create_b_0x80000.step.dependOn(&cp_config.step);
+
+                // followed by 0 filler until 0x80000
+                const dd = bash(b, "dd status=none bs=1 count=524284 if=/dev/zero >> {}", .{kernel});
+                dd.step.dependOn(&create_b_0x80000.step);
+
+                // followed finally by kernel that starts at 0x80000
+                const cat = bash(b, "cat {}/kernel8.img >> {}", .{ b.cache_root, kernel });
```
Comment (SamTebbs33) [inline build.zig]: Why is this necessary? Shouldn't it just be a case of packaging the kernel image, config etc. into a .img file?

File: build.zig [hunk 24]
```diff
@@ -67,20 +67,79 @@ pub fn build(b: *Builder) !void {
...
+            const wget = b.addSystemCommand(&[_][]const u8{
+                "wget",
+                try std.fmt.allocPrint(b.allocator, "--directory-prefix={}", .{zip_folder}),
+                "--quiet",
+                "--timestamp",
+                try std.fmt.allocPrint(b.allocator, "{}/bootcode.bin", .{firmware_url}),
+                try std.fmt.allocPrint(b.allocator, "{}/fixup.dat", .{firmware_url}),
+                try std.fmt.allocPrint(b.allocator, "{}/start.elf", .{firmware_url}),
+            });
+            wget.step.dependOn(&emulate_armstub.step);
```
Comment (SamTebbs33) [inline build.zig]: Can we store these in the repo instead of downloading them? If licensing doesn't allow this then we could request the user download them and put them in a specific place.

File: build.zig [hunk 25]
```diff
@@ -67,20 +67,79 @@ pub fn build(b: *Builder) !void {
...
+
+            const zip = b.addSystemCommand(&[_][]const u8{
+                "zip",
+                "--junk-paths",
+                "--quiet",
+                "--recurse-paths",
+                try std.fmt.allocPrint(b.allocator, "{}.zip", .{zip_folder}),
+                zip_folder,
+            });
+            zip.step.dependOn(&wget.step);
```
Comment (SamTebbs33) [inline build.zig]: Why do we need a zip now that we have the image file that we should be able to burn directly to the SD card?

File: build.zig [hunk 26]
```diff
@@ -172,3 +231,11 @@ pub fn build(b: *Builder) !void {
     });
     debug_step.dependOn(&debug_cmd.step);
 }
+
+fn bash(b: *Builder, comptime fmt: []const u8, args: anytype) *std.build.RunStep {
+    return b.addSystemCommand(&[_][]const u8{
+        "bash",
```
Comment (SamTebbs33) [inline build.zig]: As discovered by marler8997, not all systems have bash (#131). I suggest using `b.addSystemCommand` instead.

**Review by SamTebbs33 — CHANGES_REQUESTED** [review]

**Comment by SamTebbs33** [comment]
Thanks for cleaning up the image creation code in build.zig :+1:

File: build.zig (hunk 23 above)
Comment (markfirmware) [reply build.zig]: This is part of putting the image at 0x80000 with a jump instruction at 0. The bash has been replaced by zig.

https://github.com/markfirmware/pluto/blob/0fb94376e5622874046f0d874f15966c373f59fa/build.zig#L214-L242

Comment (markfirmware) [reply build.zig]: The entire disk is now packaged into rpi-sdcard.img.

Comment (markfirmware) [reply build.zig]: @SamTebbs33

**Comment by SamTebbs33** [comment]
Thanks for replacing the wget and bash :+1: Including the firmware files here means that we need to replicate their copyright notices in the README, but I can do that before finishing the port. I think there are still a few unresolved comments left then we can go ahead and merge.

File: src/kernel/arch/aarch64/rpi.zig (hunk 14 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/rpi.zig]: There were some alignment faults some time ago, but they have not recurred since enabling the mmu. I have removed this unneeded code.

File: src/kernel/arch/aarch64/arch.zig (hunk 3 above)
Comment (markfirmware) [reply src/kernel/arch/aarch64/arch.zig]: I have added comments regarding the use of uart0/uart1 in relation to qemu and also to the system clock. enable_uart=1 is now included in config.txt to ensure the baud rate is stable.

**Comment by markfirmware** [comment]
I have added the initSerial comments and removed the now unnecessary assembly code for the led.

I am now starting the /// commentary.

**Comment by markfirmware** [comment]
@SamTebbs33 Please check  997d157 for satisfaction.

**Review by SamTebbs33 — APPROVED** [review]
Thanks @markfirmware :+1:. It's great to have your changes to make it work on real hardware!
